### PR TITLE
[BREAKING] Drop constant and payable from ABI.

### LINKED ISSUE #2502: Drop constant keyword from functions

### LINKED ISSUE #4755: Drop the constant and payable keywords from the ABI
This was originally part of #2502.

Both `constant` and `payable` are replaced with `stateMutability` which has been added close to a year ago.

### PR DESCRIPTION
**Pull request by ekpyron**
Closes #4755.

Part of #3656.

To be discussed:
  - How to deal with breaking changes and the contracts in ``test/compilationTests/``?
  - Does it make sense to merge changes like this to the ``050`` branch at this point?

### CODE REVIEW

**Comment by ekpyron** [comment]
One of the reasons to create this was to see how the tests behave and start a discussion about that.
As expected, the tests involve external contracts, e.g. from truffle and the DAO, that are not yet forward compatible with changes like this, so we have to decide how to deal with this.

**Comment by axic** [comment]
@ekpyron can we merge the `constant` -> `view` changes (docs, styleguide, std, test) separately on `develop`? That it not breaking.

**Comment by ekpyron** [comment]
@axic That's true - I have them in separate commits here, so I'll extract them into a separate PR shortly!

**Comment by ekpyron** [comment]
Once #4093 is merged to develop and develop is merged into 050, I'll rebase this PR.

File: libsolidity/interface/ABI.cpp
Comment (axic) [inline libsolidity/interface/ABI.cpp:40]: We should put this warning into the documentation saying it will be removed with 0.5.0.

**Comment by axic** [comment]
The ViewPureChecker must be updated to be turned on for `view` (there's a 0.5.0 check there).

**Comment by ekpyron** [comment]
@axic The two last commits now enable enforcing ``view`` in ViewPureChecker. The last commit merely adjusts some end-to-end-tests and I extracted it to #4093 as well.
Actually I would have considered enabling these errors to be separate from this PR, but I guess it makes sense to add it in here.

**Comment by ekpyron** [comment]
Of course, since this last change (``ViewPureChecker`` reports errors for ``view``), the tests will fail again and are much more involved to update. Since we decided to disable the compilation tests for 0.5.0 I will not update them any further.

**Comment by ekpyron** [comment]
Rebased and removed the commit that updates ``test/compilationTests`` (since we decided to disable those tests for 0.5.0).

File: docs/abi-spec.rst
Comment (chriseth) [inline docs/abi-spec.rst]: Perhaps explain that `view` is the former `constant`?

File: libsolidity/analysis/ViewPureChecker.cpp
Comment (chriseth) [inline libsolidity/analysis/ViewPureChecker.cpp]: Can you use `contracts += source->filteredNodes<ContractDefinition>(source->nodes())`?

File: test/contracts/LLL_ERC20.cpp
Comment (chriseth) [inline test/contracts/LLL_ERC20.cpp]: This is lll, do we really want to change it?

File: test/libsolidity/ASTLegacyJSON.cpp
Comment (chriseth) [inline test/libsolidity/ASTLegacyJSON.cpp]: Can you add the corresponding check about stateMutability?

**Review by chriseth — DISMISSED** [review]

File: docs/abi-spec.rst
Comment (ekpyron) [reply docs/abi-spec.rst]: I'm not sure - for the abi it holds specifically that ``constant`` = (``pure`` or ``view``), so if people miss the ``constant`` field, that's what they should use instead - the fact that ``view`` replaces ``constant`` should rather be in the docs for state mutability itself and I think it's there already - I'll check.

File: libsolidity/analysis/ViewPureChecker.cpp
Comment (ekpyron) [reply libsolidity/analysis/ViewPureChecker.cpp]: Didn't know that that is defined for std::vector in libdevcore - I'll change it.

File: test/libsolidity/ASTLegacyJSON.cpp
Comment (ekpyron) [reply test/libsolidity/ASTLegacyJSON.cpp]: Done.

**Comment by ekpyron** [comment]
@chriseth Should we maybe disable the compilation tests on 050 already?

**Comment by ekpyron** [comment]
I think I now caught all errors in the in-tree compilation tests. Not sure what to do about the external tests, though.

**Comment by axic** [comment]
Changed the base and rebased.

File: docs/abi-spec.rst
Comment (axic) [inline docs/abi-spec.rst]: `stateMutability` should also be marked as a keyword (double backticks)

**Comment by axic** [comment]
Why are tests failing?

Do you want to merge the commit of removing `constant` from the local test quite before there's a need for too much rebasing?

**Comment by ekpyron** [comment]
@axic I think the test failure was due to some random RPC failure unrelated to this PR - I just rebased, so we'll see whether the tests fail again.
Otherwise I think it should be fine to merge it all at once, but I'll recheck, when I'm back in Berlin next week.

**Comment by ekpyron** [comment]
The remaining test failures should be due to the external compilation tests. I think we should merge #4206 - I created a PR with the required changes to the gnosis contracts here: https://github.com/axic/pm-contracts/pull/1 ; zeppelin seems to have removed ``constant`` already.

File: libsolidity/analysis/ViewPureChecker.cpp
Comment (axic) [inline libsolidity/analysis/ViewPureChecker.cpp]: Does this needs this local variable or otherwise the overloaded `+=` operator won't see it as a `const&`?

File: libsolidity/parsing/Parser.cpp
Comment (axic) [inline libsolidity/parsing/Parser.cpp]: Nice! I guess it makes sense to keep this at least for a while.

File: test/libsolidity/ASTLegacyJSON.cpp
Comment (axic) [inline test/libsolidity/ASTLegacyJSON.cpp]: @chriseth @ekpyron Hm, do we also want to remove this and prefer `stateMutability`? Not as part of this PR though.

File: libsolidity/analysis/ViewPureChecker.cpp
Comment (ekpyron) [reply libsolidity/analysis/ViewPureChecker.cpp]: Actually, there's no reason for the local variable - I'll remove it.

File: test/libsolidity/SolidityABIJSON.cpp
Comment (axic) [inline test/libsolidity/SolidityABIJSON.cpp]: Same question applies here.

File: test/libsolidity/ASTLegacyJSON.cpp
Comment (ekpyron) [reply test/libsolidity/ASTLegacyJSON.cpp]: Yes, I think it makes sense to remove ``payable`` as well and to always use ``stateMutability`` to avoid information duplication.

File: test/compilationTests/zeppelin/crowdsale/Crowdsale.sol
Comment (axic) [inline test/compilationTests/zeppelin/crowdsale/Crowdsale.sol]: This is non-view because of `msg.value` IIRC, but why doesn't it require `payable` then?

This might be something we need to think about a bit.

File: test/compilationTests/milestonetracker/RLP.sol
Comment (axic) [inline test/compilationTests/milestonetracker/RLP.sol]: Why is is non-view or even non-pure? It only does mstore/mload.

File: test/compilationTests/zeppelin/token/LimitedTransferToken.sol
Comment (axic) [inline test/compilationTests/zeppelin/token/LimitedTransferToken.sol]: Why is this non-view? `balanceOf` is still marked as `view`.

Comment (axic) [reply test/compilationTests/zeppelin/token/LimitedTransferToken.sol]: Hm, is it because the `interface` is declared non-view and this still doesn't let a stricter set to be set than in the interface?

File: test/compilationTests/milestonetracker/RLP.sol
Comment (ekpyron) [reply test/compilationTests/milestonetracker/RLP.sol]: As far as I remember it's the jump in the inline assembly that's not considered pure or even view at the moment (making sure that it always jumps to a piece of code that's view or pure is probably non-trivial).

File: test/compilationTests/zeppelin/token/LimitedTransferToken.sol
Comment (ekpyron) [reply test/compilationTests/zeppelin/token/LimitedTransferToken.sol]: Actually, at the moment I don't see a reason for this one and the same one in ``VestedToken.sol`` not both to be view - I'll change it and see what the tests say.

File: test/compilationTests/milestonetracker/RLP.sol
Comment (axic) [reply test/compilationTests/milestonetracker/RLP.sol]: Right, that makes sense, we should document this properly and how to avoid it: use if/for/switch.

**Comment by ekpyron** [comment]
We also need to decide how to deal with ``DAO`` and ``truffle`` tests (unless I have missed a decision there).

**Comment by axic** [comment]
Truffle has been fixed for all these things, but the projects using them aren't using the latest release yet. I've asked Gnosis, will check Zeppelin too.

As for the solc-js I'd prefer to create a DAO050 directory and keep the 0.4.x version intact because I'd like to run solc-js tests on different version as it should be the backwards compatibility glue.

**Comment by axic** [comment]
Merged the solc-js changes and updated the `solcjsTests.sh` to use it in #4280. I think that should work, be merged, and then this PR should be close to a mergeable state.

**Comment by axic** [comment]
Rebased.

**Comment by axic** [comment]
Only the two external projects (gnosis, zeppelin) are failing. Getting close.

**Comment by ekpyron** [comment]
I'm not entirely sure what to make of the 20000 line output of the still failing external tests...

**Comment by chriseth** [comment]
Requires rebase.

**Comment by axic** [comment]
Do we want to split out the (syntax) test updates?

**Comment by ekpyron** [comment]
I split this up further and extracted #4409 and #4413. What remains for this PR is the JSON ABI change, but that still breaks the openzeppelin test suite.

**Comment by ekpyron** [comment]
@axic @chriseth As far as I can tell, the external test failure is due to truffle relying on the ``constant`` field to be part of the json ABI. Not sure how to proceed (shall I look into modifying truffle? shall we open an issue at their repo? shall we keep ``constant`` in the ABI for the time being?).

**Comment by ekpyron** [comment]
There's an issue for truffle https://github.com/trufflesuite/truffle/issues/1093
and a PR for web3  https://github.com/ethereum/web3.js/pull/1765 now.

**Comment by ekpyron** [comment]
I just discussed with @chriseth that it may make sense to postpone this change for now - unless everything that depends on ``constant`` in the ABI is updated, it's likely to break a lot of existing javascript code - and only has very limited gain. We should still try to get external code updated, so that we may change this in the future.

File: docs/abi-spec.rst
Comment (axic) [inline docs/abi-spec.rst]: We should create a PR adding a deprecation warning for both `constant` and `payable`.

Comment (axic) [reply docs/abi-spec.rst]: Did that in #4549.

**Comment by axic** [comment]
This PR changes both the AST JSON and the ABI output. I think we can merge the AST JSON change, but as agreed the ABI should remain a little bit longer.

**Comment by axic** [comment]
We should also submit a PR to go-ethereum's abigen and potentially parity's abigen.

Truffle 5.x was updated and they plan to backport the changes 4.x. @frozeman made web3 releases.

@yann300 @iurimatias is remix / embark making use of `stateMutability`?

**Comment by axic** [comment]
@ekpyron can you create a PR with the AST changes only?

**Comment by ekpyron** [comment]
@axic I will.

File: docs/abi-spec.rst
Comment (axic) [inline docs/abi-spec.rst]: See  the `warning` below, which has the same content.

**Comment by iurimatias** [comment]
@axic we do and it's a breaking change for us, we'll need to update

**Comment by axic** [comment]
@chriseth @ekpyron I think we can ~~close this now~~ postpone this now. Since I consider this a "protocol change", it doesn't necessarily needs to be aligned with breaking releases. Just the time has to be right for the ecosystem to adapt.

**Comment by ekpyron** [comment]
@axic Yes, I think we were pretty much heading towards postponing this anyways and classifying this as ``protocol change`` and decoupling those from breaking releases makes sense.

**Comment by chriseth** [comment]
Removed from 0.5.0.

File: libsolidity/interface/ABI.cpp
Comment (axic) [reply libsolidity/interface/ABI.cpp:40]: Done.

**Review by axic — DISMISSED** [review]
Needs to be rebased.

**Comment by axic** [comment]
It is missing a changelog entry and all the tests are failing.

**Comment by ekpyron** [comment]
Easy to recreate and unclear when we'll actually do this, so I'm closing this PR for now and added the issue to the backlog.

**Pull request closed without merging**
